### DISCUSSION TITLE
High CPU Usage on AARCH64 mergerfs version: 2.34.1

### DISCUSSION THREAD

**Discussion by rubenkr** [trapexit/mergerfs#1135] [General]
**Describe the bug**

- Four drives (LUKS/EXT4) connected to SATA expander on Marvell Board (Espressobin) Archlinux ARM64
- Drives mounted to /mnt*
- Shared with NFS
- Mergerfs Mount /etc/fstab as follows:
`/mnt/* /storage fuse.mergerfs allow_other,cache.files=partial,use_ino,dropcacheonclose=true,link_cow=true,category.create=mfs,cache.entry=120,cache.attr=120,cache.statfs=10,minfreespace=20G,fsname=mergerfsPool 0 0
`

**General Information on setup**
OS: Arch Linux ARM aarch64
Host: Globalscale Marvell ESPRESSOBin Board
Kernel: 5.10.16-1-ARCH
CPU: (2) @ 800MHz
Memory: 209MiB / 988MiB
mergerfs version: 2.34.1
Disks: all are ext4 formatted and LUKS encrypted

➜  ~ mergerfs -V
`mergerfs version: 2.34.1
`

➜  ~ cat /proc/cpuinfo
```
processor       : 0
BogoMIPS        : 25.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 1
BogoMIPS        : 25.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4
```

➜  ~ df -h
```
Filesystem                Size  Used Avail Use% Mounted on
dev                       458M     0  458M   0% /dev
run                       495M  444K  494M   1% /run
/dev/mmcblk0p1             29G  8.2G   20G  30% /
tmpfs                     495M     0  495M   0% /dev/shm
tmpfs                     495M  4.0K  495M   1% /tmp
tmpfs                      99M     0   99M   0% /run/user/1001
/dev/sde1                 2.7T  1.9T  764G  71% /mnt/3TB
/dev/mapper/WD14TB         13T   12T  773G  95% /mnt/WD14TB
/dev/mapper/WD10TB-0EZAZ  9.1T  8.3T  790G  92% /mnt/WD10TB-0EZAZ
/dev/mapper/WD10TB-1EDBZ  9.1T  1.2T  7.9T  13% /mnt/WD10TB-1EDBZ
mergerfsPool               34T   24T   11T  70% /storage
```

➜  ~ lsblk -f
```
NAME           FSTYPE      FSVER LABEL UUID                                 FSAVAIL FSUSE% MOUNTPOINTS
sda            crypto_LUKS 2           faf88f52-2fe3-422d-87fb-2e1190e571ff
`-WD8TB
sdb            crypto_LUKS 2           6ca3f142-35b7-4eff-acfc-28357337aae1
`-WD10TB-1EDBZ ext4        1.0         e69bab8c-a817-4d7d-ba7a-7daf11096ce5    7.9T    13% /mnt/WD10TB-1EDBZ
sdc            crypto_LUKS 2           9e2c97a8-6e89-4a07-940a-3a21694711f7
`-WD14TB       ext4        1.0         da228486-e148-4407-8080-afa95f548cc5  772.5G    94% /mnt/WD14TB
sdd            crypto_LUKS 2           cf5f4dbb-c1eb-4316-8c0c-c35530943c56
`-WD10TB-0EZAZ ext4        1.0         ee826141-3670-457f-a74a-2683da0fc032  789.2G    91% /mnt/WD10TB-0EZAZ
sde
`-sde1         ext4        1.0         b0c4160e-461a-4885-9065-b085b683401c  763.7G    67% /mnt/3TB
mmcblk0
`-mmcblk0p1    ext4        1.0         0898b992-177f-4535-b216-ee4d7dadb2ff   19.3G    28% /
```

write benchmark on individual drives
```
$ dd if=/dev/zero of=/mnt/3TB/1GB.file bs=1M count=1024 oflag=nocache conv=fdatasync status=progress
535822336 bytes (536 MB, 511 MiB) copied, 14 s, 38.3 MB/s

dd if=/dev/zero of=/mnt/WD14TB/1GB.file bs=1M count=1024 oflag=nocache conv=fdatasync status=progress
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 37.2189 s, 28.8 MB/s
```

**➜  /mnt ls -l**
`
drwxrwxrwx 15 plex plex 4096 Jan 29 10:42 3TB
drwxrwxrwx 18 plex plex 4096 Jan 29 10:40 WD10TB
drwxrwxrwx 25 plex plex 4096 Jan 29 11:40 WD14TB
drwxrwxrwx 11 plex plex 4096 Jan 29 11:41 WD8TB
`
write benchmark mergerfspool
`dd if=/dev/zero of=/storage/1GB.file bs=1M count=1024 oflag=nocache conv=fdatasync status=progress
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 44.4486 s, 24.2 MB/s

High CPU Usage
![image](https://user-images.githubusercontent.com/1043739/216724860-9c07b506-b21d-43bf-809e-fc01a01644a9.png)

**To Reproduce**

- read/write operation (e.g. copying via nfs share) maxing out CPU at 100%

**Expected behavior**

- cpu not running onto full 100%
- potentially there is a ARM specific setting I have not found in the github Read.me?

Any idea to further investigate is much appreciated knowing the fact that ARM systems are really bitchy to debug in that regards -_-

**Comment by trapexit** [comment]
It might just be the cost of doing business. That looks like a pretty low end cpu (by modern standards) and using LUKS and NFS and mergerfs just could be too much for it. The best thing to test would be to remove different components and compare. Check how much CPU dd is taking when reading/writing just to the LUKS device. Try mergerfs over a single non LUKS device. Etc. Maybe try using a single thread and see if perf changes much. I'm going to be adding some options to pin threads to cores that might help.

Unfortunately, I don't have any systems with that chipset. I have a XU4 and a La Frite SBC I could try setting up similarly to test but there really is too much that can be done to impact what mergerfs has to do. Enabling all the different caches can help.

**Comment by rubenkr** [comment]
Awesome @trapexit a separate thread would be much of help. As of know I am impressed and inspired by one of your real life examples ;). So long I will close this by now as it is too broad of being a bug itself. Thank you so much for your work and passion.

**Reply by trapexit** [reply]
Not sure what you mean by "a separate thread". mergerfs is already multithreaded and you can control the value via the `threads` argument. In a future release there will be more control but my point was trying to limit the number of threads or where they run.

**Reply by rubenkr** [reply]
Thread was meant not to be technical but as a written documentation here on github. Sorry for confusion.

**Reply by trapexit** [reply]
Ah. What do you mean by that? Document what?

**Reply by rubenkr** [reply]
You wrote 
> I'm going to be adding some options to pin threads to cores that might help.

and
> Let's move this to a discussion and if anything comes up that can be improved or just more information it can still be published. Can always open new issue tickets if something is found.


in that [comment](https://github.com/trapexit/mergerfs/discussions/1135#discussioncomment-4870200) of yours. Maybe I am lost in translation but I had the hope that any change for adaptions towards aarch64 would be communicated in the README or somewhere else.

**Reply by trapexit** [reply]
The thread pinning will be documented naturally but there is nothing special for aarch64. Nothing about the project is architecture specific.

**Comment by trapexit** [comment]
Let's move this to a discussion and if anything comes up that can be improved or just more information it can still be published. Can always open new issue tickets if something is found.
